### PR TITLE
Shard download space fix

### PR DESCRIPTION
**Pull request by undertome**
## High Level Overview of Change

With this change, the `DatabaseShardImp` will confirm that sufficient space is available to store _all_ the shards queued for download, instead of confirming that each shard will fit without considering the additional space required by the others.

### Context of Change

This PR is stacked on two unmerged PRs. The relevant code starts with commit 6e1b74e.

### Type of Change

- [ X] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by scottschurr** [comment]
Hi @undertome, it would be useful for me to have a little more background on this bug fix.  It would be handy to know:

- The user observed nature of the bug before the fix.
- The nature of the improved user experience after the bug fix.

Also it looks like the fix is associated with being low on resources.  Does the user have any recourse once they discover that they are too low on resources to handle the shards?

Thanks.

**Comment by undertome** [comment]
> Hi @undertome, it would be useful for me to have a little more background on this bug fix. It would be handy to know:
> 
>     * The user observed nature of the bug before the fix.
> 
>     * The nature of the improved user experience after the bug fix.
> 
> 
> Also it looks like the fix is associated with being low on resources. Does the user have any recourse once they discover that they are too low on resources to handle the shards?
> 
> Thanks.

Hi Scott,

- Before the fix, the user would be able to initiate a batch of shard downloads when there was insufficient space to store all the shards, but sufficient space to store each shard in isolation. The user would see the first download complete, but subsequent downloads would fail.

- With the fix, a request to download a batch of shards will fail immediately if there is insufficient space to store all the shards and a descriptive log message will be created.

If the user is low on resources, the `DatabaseShardImp` will log a warning, discontinue acquiring shards, and reject requests to download shards. At this point, it would be the user's responsibility to provide more storage to accommodate additional shards. 

Hope that helps.

**Comment by scottschurr** [comment]
A couple of your commit messages don't follow these guidelines: https://chris.beams.io/posts/git-commit/  May I suggest rewordings?  Consider:
```
Delay SSLHTTPDownloader dtor until current session finishes
```
and
```
Download queued shards only if there is space for them all
```
Or perhaps you can come up with something better.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 1]
```diff
@@ -303,62 +303,103 @@ DatabaseShardImp::prepareLedger(std::uint32_t validLedgerSeq)
...
+        };
+        if (!seqCheck(app_.getLedgerMaster().getValidLedgerIndex() + 1) ||
+            !seqCheck(app_.getLedgerMaster().getCurrentLedgerIndex()))
+        {
+            return false;
+        }
+
+        if (shards_.find(shardIndex) != shards_.end())
+        {
+            JLOG(j_.debug()) << "shard " << shardIndex
```
Comment (scottschurr) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:362]: FWIW, not exercised in unit tests.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 2]
```diff
@@ -303,62 +303,103 @@ DatabaseShardImp::prepareLedger(std::uint32_t validLedgerSeq)
...
-            return fail("has an invalid index");
-        return true;
-    };
-    if (!seqCheck(app_.getLedgerMaster().getValidLedgerIndex() + 1) ||
-        !seqCheck(app_.getLedgerMaster().getCurrentLedgerIndex()))
+    auto const numHistShards = numHistoricalShards(lock);
+
+    // Check shard count and available storage space
+    if (numHistShards + historicalShardsToPrepare > maxHistoricalShards_)
+        return fail("maximum number of historical shards reached");
```
Comment (scottschurr) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:377]: Not exercised in unit tests.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 3]
```diff
@@ -303,62 +303,103 @@ DatabaseShardImp::prepareLedger(std::uint32_t validLedgerSeq)
...
+
+    if (historicalShardsToPrepare)
     {
-        return false;
+        // Check available storage space for historical shards
+        if (!sufficientStorage(
+                historicalShardsToPrepare, PathDesignation::historical, lock))
+            return fail("insufficient storage space available");
```
Comment (scottschurr) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:384]: Not exercised in unit tests.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 4]
```diff
@@ -303,62 +303,103 @@ DatabaseShardImp::prepareLedger(std::uint32_t validLedgerSeq)
...
+            shardIndexes.size() - historicalShardsToPrepare;
+        recentShardsToPrepare)
     {
-        JLOG(j_.debug()) << "shard " << shardIndex
-                         << " is already stored or queued for import";
-        return false;
+        // Check available storage space for recent shards
+        if (!sufficientStorage(
+                recentShardsToPrepare, PathDesignation::none, lock))
+            return fail("insufficient storage space available");
```
Comment (scottschurr) [inline src/ripple/nodestore/impl/DatabaseShardImp.cpp:394]: Not exercised in unit tests.

**Review by scottschurr — APPROVED** [review]
:+1: Looks good.  I left notes about a few things for you to think about.  Nothing critical, although covering a few more of the error cases in the unit tests would be really nice.  All comments are at your discretion.  Thanks for making life a little easier for ledger users!

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 5]
```diff
@@ -303,62 +303,103 @@ DatabaseShardImp::prepareLedger(std::uint32_t validLedgerSeq)
...
+        std::transform(
+            shardIndexes.begin(),
+            shardIndexes.end(),
+            indexesAsString.begin(),
+            [](uint32_t const index) { return std::to_string(index); });
+
+        std::string const prequel = shardIndex
+            ? "shard " + std::to_string(*shardIndex)
+            : std::string("shard") + (shardIndexes.size() > 1 ? "s " : " ") +
+                boost::algorithm::join(indexesAsString, ", ");
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:321]: Done

File: src/ripple/net/impl/HTTPStream.cpp [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+{
+    assert(stream_);
+    return stream_->next_layer();
+}
+
+bool
+SSLStream::connect(
+    std::string& errorOut,
+    std::string const host,
+    std::string const port,
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:44]: `host` and `port` can be const references.

File: src/ripple/net/impl/HTTPStream.cpp [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+{
+    assert(stream_);
+    return *stream_;
+}
+
+bool
+RawStream::connect(
+    std::string& errorOut,
+    std::string const host,
+    std::string const port,
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:133]: `host` and `port` can be const references.

File: src/ripple/net/HTTPStream.h [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    [[nodiscard]] virtual boost::asio::ip::tcp::socket&
+    getStream() = 0;
+
+    [[nodiscard]] virtual bool
+    connect(
+        std::string& errorOut,
+        std::string const host,
+        std::string const port,
```
Comment (miguelportilla) [inline src/ripple/net/HTTPStream.h:58]: `host` and `port` can be const references.

File: src/ripple/net/HTTPStream.h [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    virtual ~SSLStream() = default;
+
+    boost::asio::ip::tcp::socket&
+    getStream() override;
+
+    bool
+    connect(
+        std::string& errorOut,
+        std::string const host,
+        std::string const port,
```
Comment (miguelportilla) [inline src/ripple/net/HTTPStream.h:93]: `host` and `port` can be const references.

File: src/ripple/net/HTTPStream.h [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+    virtual ~RawStream() = default;
+
+    boost::asio::ip::tcp::socket&
+    getStream() override;
+
+    bool
+    connect(
+        std::string& errorOut,
+        std::string const host,
+        std::string const port,
```
Comment (miguelportilla) [inline src/ripple/net/HTTPStream.h:134]: `host` and `port` can be const references.

File: src/ripple/net/impl/HTTPStream.cpp [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+    assert(stream_);
+    return stream_->next_layer();
+}
+
+bool
+SSLStream::connect(
+    std::string& errorOut,
+    std::string const host,
+    std::string const port,
+    boost::asio::yield_context& yield)
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:45]: We've lost `ec` error message loging. Consider passing in an `ec` ref as done in the other functions.

File: src/ripple/net/impl/HTTPDownloader.cpp [hunk 12]
```diff
@@ -161,36 +166,12 @@ SSLHTTPDownloader::do_session(
     // connection:
     std::uint64_t const rangeStart = size(p);
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPDownloader.cpp:155]: This local should be declared after the `connect` call.

File: src/ripple/net/HTTPStream.h [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    boost::asio::io_service::strand& strand_;
+};
+
+class RawStream : public HTTPStream
+{
+public:
+    RawStream(
+        Config const& config,
+        boost::asio::io_service::strand& strand,
+        beast::Journal j);
```
Comment (miguelportilla) [inline src/ripple/net/HTTPStream.h:123]: `config` and `journal` are unused. These variable should be remove from the ctor. I recommend eliminating the `makeUnique` and modifying `HTTPDownloader` as such.
eg.
```
    if (ssl)
        stream_ = std::make_unique<SSLStream>(config_, strand_, j_);
    else
        stream_ = std::make_unique<RawStream>(strand_);
```

File: src/ripple/net/impl/HTTPStream.cpp [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    using namespace boost::beast;
+
+    boost::system::error_code ec;
+
+    auto fail = [&errorOut](std::string const& errorIn) {
+        errorOut = errorIn;
+        return false;
+    };
+
+    ip::tcp::resolver resolver{strand_.context()};
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:57]: Can this ctor throw?

File: src/ripple/net/impl/HTTPStream.cpp [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:146]: Can this ctor throw?

File: src/ripple/net/impl/HTTPStream.cpp [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+    parser& p,
+    bool readSome,
+    boost::asio::yield_context& yield,
+    boost::system::error_code& ec)
+{
+    if (readSome)
+        boost::beast::http::async_read_some(*stream_, buf, p, yield[ec]);
+    else
+        boost::beast::http::async_read(*stream_, buf, p, yield[ec]);
+}
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:189]: Consider adding `HTTPStream::ayncReadSome` and removing the `::asyncRead` `readSome` params.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 17]
```diff
@@ -303,62 +303,103 @@ DatabaseShardImp::prepareLedger(std::uint32_t validLedgerSeq)
...
+    auto historicalShardsToPrepare = 0;
 
-    if (shardIndex < earliestShardIndex())
+    for (auto const shardIndex : shardIndexes)
     {
-        return fail(
-            "comes before earliest shard index " +
-            std::to_string(earliestShardIndex()));
+        if (!canAdd_)
+            return fail("cannot be stored at this time");
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:335]: Added unit tests to exercise additional paths

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 1 above)
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:362]: Added unit tests to exercise additional paths

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 2 above)
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:377]: Added unit tests to exercise additional paths

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 3 above)
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:384]: Added unit tests to exercise additional paths, but this specific error condition is not triggered.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp (hunk 4 above)
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:394]: Added unit tests to exercise additional paths, but this specific error condition is not triggered.

File: src/ripple/nodestore/impl/DatabaseShardImp.cpp [hunk 18]
```diff
@@ -303,62 +303,103 @@ DatabaseShardImp::prepareLedger(std::uint32_t validLedgerSeq)
...
+            return fail(
+                "comes before earliest shard index " +
+                    std::to_string(earliestShardIndex()),
+                shardIndex);
+        }
+
+        // If we are synced to the network, check if the shard index
+        // is greater or equal to the current shard.
+        auto seqCheck = [&](std::uint32_t seq) {
+            // seq will be greater than zero if valid
```
[the remark this answers is not in the document]
Comment (undertome) [reply src/ripple/nodestore/impl/DatabaseShardImp.cpp:348]: Lines 345 and 346 still seem relevant to me - but maybe line 348 should say ">= earliest ledger sequence".

**Comment by scottschurr** [comment]
FWIW, I was running the unit tests for code coverage on this branch using macOS and I got this error message during the test run:
```
ERR:ShardArchiveHandler async_connect
ERR:ShardArchiveHandler Downloading shard id 10 from URL 127.0.0.1/10.tar.lz4
rippled(2843,0x13c67c5c0) malloc: *** error for object 0x13eed4000: pointer being freed was not allocated
rippled(2843,0x13c67c5c0) malloc: *** set a breakpoint in malloc_error_break to debug
make[3]: *** [CMakeFiles/coverage_report] Abort trap: 6
make[2]: *** [CMakeFiles/coverage_report.dir/all] Error 2
make[1]: *** [CMakeFiles/coverage_report.dir/rule] Error 2
make: *** [coverage_report] Error 2
```
It did not leave a stack trace behind.  I'll see if I can reproduce the problem.

**Comment by scottschurr** [comment]
Not surprisingly the bug is not easy to reproduce.

- Running the unit tests alone (once) did not recreate it.
- Running code coverage on just `ShardArchiveHandler` did not reproduce it.
- Running all of the unit tests under code coverage again did not reproduce it.

I'm going to start a loop running the `ShardArchiveHandler` test to see if I can reproduce the bug with enough repetitions.

**Review by scottschurr — APPROVED** [review]
Thanks for the improved unit test coverage.

I'm guessing the non-reproducible unit test failure that I saw is not because of these changes.

**Comment by scottschurr** [comment]
I continued running a test over the weekend to try and get more information about the crash observed on Friday.  I ran just the ShardArchiveHandler unit test in a loop.  After 5928 iterations I got a different crash from the originally reported crash.  Here's link to the stack trace of this most recent crash in case it's useful: https://gist.github.com/scottschurr/cec16dc035d9ec83fe61e6411cee0ddf

**Comment by scottschurr** [comment]
Success!  By running just the ShardArchiveHandler under code coverage in a loop I was able to reproduce the original symptom quickly.  And this time I got a stack trace.  On the console I see:
```
ERR:ShardArchiveHandler Downloading shard id 9 from URL 127.0.0.1/9.tar.lz4
ERR:ShardArchiveHandler async_connect
ERR:ShardArchiveHandler Downloading shard id 10 from URL 127.0.0.1/10.tar.lz4
rippled(2685,0x148d915c0) malloc: *** error for object 0x13f548000: pointer being freed was not allocated
rippled(2685,0x148d915c0) malloc: *** set a breakpoint in malloc_error_break to debug
make[3]: *** [CMakeFiles/coverage_report] Abort trap: 6
make[2]: *** [CMakeFiles/coverage_report.dir/all] Error 2
make[1]: *** [CMakeFiles/coverage_report.dir/rule] Error 2
make: *** [coverage_report] Error 2
```
You can find the stack trace here: https://gist.github.com/scottschurr/f43be7a61290562738728bb4b668a6e8

**Comment by undertome** [comment]
> I continued running a test over the weekend to try and get more information about the crash observed on Friday. I ran just the ShardArchiveHandler unit test in a loop. After 5928 iterations I got a different crash from the originally reported crash. Here's link to the stack trace of this most recent crash in case it's useful: https://gist.github.com/scottschurr/cec16dc035d9ec83fe61e6411cee0ddf

Are you sure you're running code from this branch? That stack trace references `SSLHTTPDownloader` which no longer exists.

**Comment by scottschurr** [comment]
> Are you sure you're running code from this branch? That stack trace references SSLHTTPDownloader which no longer exists.

That's a fair question.  Here's the git log of the branch I've been testing:

```
commit fb1ca671319fcfdbeb3943f5d1fc554c3d7d230c (HEAD -> devon-shard-download-space-fix, devon/shard-download-space-fix)
Author: Devon White <dwhite@ripple.com>
Date:   Thu Sep 17 17:26:02 2020 -0400

    [FOLD] Clang format patch

commit 09827143f62683907745a23842f6876d3612ade4
Author: Devon White <dwhite@ripple.com>
Date:   Thu Sep 17 17:21:28 2020 -0400

    [FOLD] Add unit tests

commit f9b9647a5bd6b47133f2e58e1eb6585944c2d0fa
Author: Devon White <dwhite@ripple.com>
Date:   Wed Sep 9 17:15:13 2020 -0400

    Download queued shards only if there is space for them all

commit e31c02ba89b6b7727eb2a8246d3d4ad4f227b104
Author: Devon White <dwhite@ripple.com>
Date:   Fri Aug 14 15:45:56 2020 -0400

    Delay HTTPDownloader dtor until current session
    finishes

commit 649d4a741e112777575e9f9d10100c6588037d08
Author: Devon White <dwhite@ripple.com>
Date:   Tue Aug 11 12:10:10 2020 -0400

    Allow shard downloads via HTTP or HTTPS

commit a8d481c2a54267d80bd539003f56481424b9ae63 (upstream/develop, origin/develop, develop)
Author: Nik Bougalis <nikb@bougalis.net>
Date:   Tue Sep 1 10:43:26 2020 -0700

    Set version to 1.7.0-b1
```

**Comment by scottschurr** [comment]
That said, even though that's the source code currently in place, it's possible (but seems unlikely) that I did not do a proper build before running the ShardArchiveHandler unit test in a loop over the weekend.  If that's the case please accept my apologies.

For the `malloc` failure under Codecov, however, I'm pretty confident that I have both the correct sources and that those are the source files I'm running.  The rebuild is part of how Codecov is invoked.

**Comment by undertome** [comment]
> That said, even though that's the source code currently in place, it's possible (but seems unlikely) that I did not do a proper build before running the ShardArchiveHandler unit test in a loop over the weekend. If that's the case please accept my apologies.
> 
> For the `malloc` failure under Codecov, however, I'm pretty confident that I have both the correct sources and that those are the source files I'm running. The rebuild is part of how Codecov is invoked.

The `malloc` failure seems current.

File: src/ripple/net/HTTPStream.h (hunk 8 above)
Comment (undertome) [reply src/ripple/net/HTTPStream.h:58]: Done

File: src/ripple/net/HTTPStream.h (hunk 9 above)
Comment (undertome) [reply src/ripple/net/HTTPStream.h:93]: Done

File: src/ripple/net/HTTPStream.h (hunk 10 above)
Comment (undertome) [reply src/ripple/net/HTTPStream.h:134]: Done

File: src/ripple/net/impl/HTTPDownloader.cpp (hunk 12 above)
Comment (undertome) [reply src/ripple/net/impl/HTTPDownloader.cpp:155]: Done

File: src/ripple/net/impl/HTTPStream.cpp (hunk 11 above)
Comment (undertome) [reply src/ripple/net/impl/HTTPStream.cpp:45]: Passing the error code to the write and read functions is easy because in each case, there's only one function that could be the cause of an error, whereas in the connect functions, there are many functions that could cause an error. So to provide info to the caller about which function errored, we'd have to provide additional context with the `ec` parameter. Instead of that I'd rather just provide the context (`errorOut`) without the additional `ec` parameter.

File: src/ripple/net/HTTPStream.h (hunk 13 above)
Comment (undertome) [reply src/ripple/net/HTTPStream.h:123]: Done

**Comment by miguelportilla** [comment]
I took a quick look at the stack @scottschurr produced and observed the following. The crash comes from within a journal. From `ShardArchiveHandler_test::testRedundantShardFailure`, on line 612, the test has only created `Env` called `close()` which calls `RPCCall::fromNetwork`. A possible culprit I see is that inside `fromNetwork` a local journal `j` is declared and passed into the callbacks and the HTTPClient object. This might be a problem as `HTTPClient::request` is non blocking and `j`'s sink object may have gone out of scope when the callbacks or `invokeComplete` are executed. Its possible the sink may be still be valid at this point as it might be kept alive by someone but I am not sure.

**Comment by undertome** [comment]
> I took a quick look at the stack @scottschurr produced and observed the following. The crash comes from within a journal. From `ShardArchiveHandler_test::testRedundantShardFailure`, on line 612, the test has only created `Env` called `close()` which calls `RPCCall::fromNetwork`. A possible culprit I see is that inside `fromNetwork` a local journal `j` is declared and passed into the callbacks and the HTTPClient object. This might be a problem as `HTTPClient::request` is non blocking and `j`'s sink object may have gone out of scope when the callbacks or `invokeComplete` are executed. Its possible the sink may be still be valid at this point as it might be kept alive by someone but I am not sure.

Thanks for looking into it, I will see if this checks out.

File: src/ripple/net/impl/HTTPStream.cpp (hunk 11 above)
Comment (miguelportilla) [reply src/ripple/net/impl/HTTPStream.cpp:45]: While reporting "which function errored" may be helpful, we are throwing away the cause. Typically if the `ec` is set, further progress on the process is aborted, the call stack is unwind and the `ec` is floated up. If we choose to keep things as they, it may be helpful to append its message into `errorOut`. 

eg.
```return fail("async_resolve. Error: " + ec.message());```

As is, if any of these functions fail, we don't know why.

Comment (miguelportilla) [reply src/ripple/net/impl/HTTPStream.cpp:45]: For example, if the call to `resolver.async_resolve` fails, `errorOut` will contain "async_resolve" but the reason it failed will be discarded and we will have no idea why it failed.

Comment (undertome) [reply src/ripple/net/impl/HTTPStream.cpp:45]: Done

File: src/ripple/net/impl/HTTPStream.cpp [hunk 19]
```diff
@@ -65,7 +65,7 @@ SSLStream::connect(
     }
     catch (std::exception const& e)
     {
-        return fail(std::string("exception: ") + e.what());
+        return fail("exception");
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:68]: If `ec` isn't set, we'd end up with an error message that says "exception".  I recommend keeping the exception message.

File: src/ripple/net/impl/HTTPStream.cpp [hunk 20]
```diff
@@ -156,7 +156,7 @@ RawStream::connect(
...
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:159]: See prior comment.

File: src/ripple/net/impl/HTTPStream.cpp [hunk 21]
```diff
@@ -49,8 +49,8 @@ SSLStream::connect(
 
     boost::system::error_code ec;
 
-    auto fail = [&errorOut](std::string const& errorIn) {
-        errorOut = errorIn;
+    auto fail = [&errorOut, &ec](std::string const& errorIn) {
+        errorOut = errorIn + ": " + ec.message();
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:53]: If ec is not set, we append ": ". It may be a good ide to check ec before appending it.
eg.
```
        errorOut = errorIn;
        if (ec)
            errorOut += ". Error code: " + ec.message();
```

File: src/ripple/net/impl/HTTPStream.cpp [hunk 22]
```diff
@@ -140,8 +140,8 @@ RawStream::connect(
...
```
Comment (miguelportilla) [inline src/ripple/net/impl/HTTPStream.cpp:144]: See prior comment.

File: src/ripple/net/impl/HTTPStream.cpp (hunk 19 above)
Comment (undertome) [reply src/ripple/net/impl/HTTPStream.cpp:68]: Done

File: src/ripple/net/impl/HTTPStream.cpp (hunk 20 above)
Comment (undertome) [reply src/ripple/net/impl/HTTPStream.cpp:159]: Done

File: src/ripple/net/impl/HTTPStream.cpp (hunk 21 above)
Comment (undertome) [reply src/ripple/net/impl/HTTPStream.cpp:53]: Got this done with an additional defaulted parameter to the lambda

File: src/ripple/net/impl/HTTPStream.cpp (hunk 22 above)
Comment (undertome) [reply src/ripple/net/impl/HTTPStream.cpp:144]: Done

**Review by miguelportilla — APPROVED** [review]
👍

**Comment by undertome** [comment]
New commit fixes a Travis error caused by an "unused" (it's used in an `assert` statement) variable. There was also a bug I discovered that was caused by the `Checkpointer` outliving a `DatabaseCon`.

**Review by miguelportilla — APPROVED** [review]
